src/curve-fed/CurvePoolAdapter.sol

```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.10;
+
+import "src/interfaces/IERC20.sol";
+import "src/interfaces/curve/IMetaPool.sol";
+
+abstract contract CurvePoolAdapter {
+
+    IERC20 public constant dola = IERC20(0x865377367054516e17014CcdED1e7d814EDC9ce4);
+    IMetaPool public immutable crvMetapool;
+    uint public constant PRECISION = 10_000;
+    uint public immutable CRVPRECISION = 10**18;
+
+    constructor(address crvMetapool_){
+        crvMetapool = IMetaPool(crvMetapool_);
+        //Approve max uint256 spend for crvMetapool, from this address
+        dola.approve(crvMetapool_, type(uint256).max);
+        IERC20(crvMetapool_).approve(crvMetapool_, type(uint256).max);
+    }
+    /**
+     * @notice Function for depositing into curve metapool.
+     * @param amountDola Amount of dola to be deposited into metapool
+     * @param allowedSlippage Max allowed slippage. 1 = 0.01%
+     * @return Amount of Dola-Metapool tokens bought
+     */
+    function metapoolDeposit(uint256 amountDola, uint allowedSlippage) internal returns(uint256){
+        uint[2] memory amounts = [amountDola, 0];
+        uint expectedCrvLp = dolaToLp(amountDola);
+        uint minCrvLPOut = applySlippage(expectedCrvLp, allowedSlippage);
+        return crvMetapool.add_liquidity(amounts, minCrvLPOut);
+    }
+
+    /**
+     * @notice Function for depositing into curve metapool.
+     * @param amountDola Amount of dola to be withdrawn from the metapool
+     * @param allowedSlippage Max allowed slippage. 1 = 0.01%
+     * @return Amount of Dola tokens received
+     */
+    function metapoolWithdraw(uint amountDola, uint allowedSlippage) internal returns(uint256){
+        uint[2] memory amounts = [amountDola, 0];
+        uint amountCrvLp = crvMetapool.calc_token_amount( amounts, false);
+        uint expectedCrvLp = dolaToLp(amountDola);
+        //The expectedCrvLp must be higher or equal than the crvLp amount we supply - the allowed slippage
+        require(expectedCrvLp >= applySlippage(amountCrvLp, allowedSlippage), "LOSS EXCEED WITHDRAW MAX LOSS");
+        uint dolaMinOut = applySlippage(amountDola, allowedSlippage);
+        return crvMetapool.remove_liquidity_one_coin(amountCrvLp, 0, dolaMinOut);
+    }
```

Comment:
No defense against sandwich attacks if DOLA has upwardly depegged or the withdraw is large enough to upwardly depeg DOLA.